rhp/v4/rpc.go

```diff
@@ -0,0 +1,705 @@
+	RPCSectorRootsResult struct {
+		Revision types.V2FileContract `json:"revision"`
+		Roots    []types.Hash256      `json:"roots"`
+		Cost     types.Currency       `json:"cost"`
+	}
+
+	// RPCFormContractResult contains the result of executing the form contract RPC.
+	RPCFormContractResult struct {
+		Contract     ContractRevision `json:"contract"`
+		FormationSet TransactionSet   `json:"formationSet"`
+		Cost         types.Currency   `json:"cost"`
+	}
+
+	// RPCRenewContractResult contains the result of executing the renew contract RPC.
+	RPCRenewContractResult struct {
+		Contract   ContractRevision `json:"contract"`
+		RenewalSet TransactionSet   `json:"renewalSet"`
+		Cost       types.Currency   `json:"cost"`
+	}
+)
+
+func callSingleRoundtripRPC(ctx context.Context, t TransportClient, rpcID types.Specifier, req, resp rhp4.Object) error {
+	s := t.DialStream(ctx)
+	defer s.Close()
+
+	if err := rhp4.WriteRequest(s, rpcID, req); err != nil {
+		return fmt.Errorf("failed to write request: %w", err)
+	} else if err := rhp4.ReadResponse(s, resp); err != nil {
+		return fmt.Errorf("failed to read response: %w", err)
+	}
+	return nil
+}
+
+// RPCSettings returns the current settings of the host.
+func RPCSettings(ctx context.Context, t TransportClient) (rhp4.HostSettings, error) {
+	var resp rhp4.RPCSettingsResponse
+	err := callSingleRoundtripRPC(ctx, t, rhp4.RPCSettingsID, nil, &resp)
+	return resp.Settings, err
+}
+
+// RPCReadSector reads a sector from the host.
+func RPCReadSector(ctx context.Context, t TransportClient, prices rhp4.HostPrices, token rhp4.AccountToken, w io.Writer, root types.Hash256, offset, length uint64) (RPCReadSectorResult, error) {
+	req := &rhp4.RPCReadSectorRequest{
+		Prices: prices,
+		Token:  token,
+		Root:   root,
+		Offset: offset,
+		Length: length,
+	}
+	if err := req.Validate(t.PeerKey()); err != nil {
+		return RPCReadSectorResult{}, fmt.Errorf("invalid request: %w", err)
+	}
+
+	s := t.DialStream(ctx)
+	defer s.Close()
+
+	if err := rhp4.WriteRequest(s, rhp4.RPCReadSectorID, req); err != nil {
+		return RPCReadSectorResult{}, fmt.Errorf("failed to write request: %w", err)
+	}
+
+	var resp rhp4.RPCReadSectorStreamedResponse
+	if err := rhp4.ReadResponse(s, &resp); err != nil {
+		return RPCReadSectorResult{}, fmt.Errorf("failed to read response: %w", err)
+	}
+
+	// TODO: verify proof
+	n, err := io.Copy(w, io.LimitReader(s, int64(resp.DataLength)))
+	if err != nil {
+		return RPCReadSectorResult{}, fmt.Errorf("failed to read data: %w", err)
+	} else if n != int64(resp.DataLength) {
+		return RPCReadSectorResult{}, io.ErrUnexpectedEOF
+	}
+	return RPCReadSectorResult{
+		Cost: prices.RPCReadSectorCost(length),
+	}, nil
+}
+
+// RPCWriteSector writes a sector to the host.
+func RPCWriteSector(ctx context.Context, t TransportClient, prices rhp4.HostPrices, token rhp4.AccountToken, rl ReaderLen, duration uint64) (RPCWriteSectorResult, error) {
```

Comment:
What's the benefit of the `ReaderLen` interface over passing in the amount of data to read?